### PR TITLE
Add interpreter for BatchNormGradOp

### LINKED ISSUE #1121: Add interpreter for BatchNormGradOp
Following the [testing guidelines](https://github.com/openxla/stablehlo/blob/main/docs/interpreter_checklist.md)

### PR DESCRIPTION
**Pull request by ghpvnist**
We have the following constraints in the spec:

```
(I1) `operand`: tensor of floating-point type.
(I2) `scale`: 1-dimensional tensor of floating-point type.
(I3) `mean`: 1-dimensional tensor of floating-point type.
(I4) `variance`: 1-dimensional tensor of floating-point type.
(I5) `grad_output`: tensor of floating-point type.
(I6) `epsilon`: constant of type `f32`.
(I7) `feature_index`: constant of type `si64`.
(C1) 0 <= `feature_index` < rank(`operand`).
(C2) `operand`, `scale`, `mean`, `variance`, `grad_output`, `grad_operand`
     `grad_scale` and `grad_offset` have the same element type.
(C3) `operand`, `grad_output` and `grad_operand` have the same shape.
(C4) `scale`, `mean`, `variance`, `grad_scale` and `grad_offset` have the
     same shape.
(C5) size(`scale`) = `dim(operand, feature_index)`.
```

These constraints will be comprehensively covered by the following tests:

```
I1: a) `operand` is not a tensor of floating-point type. (Covered by ODS).
I2: a) `scale` is not a 1-dimensional tensor. (Covered by ODS).
    b) `scale` is not a tensor of floating-point type. (Covered by ODS).
I3: a) `mean` is not a 1-dimensional tensor. (Covered by ODS).
    b) `mean` is not a tensor of floating-point type. (Covered by ODS).
I4: a) `variance` is not a 1-dimensional tensor. (Covered by ODS).
    b) `variance` is not a tensor of floating-point type. (Covered by ODS).
I5: a) `grad_output` is not a tensor of floating-point type. (Covered by ODS).
I6: a) `epsilon` is not a constant of type `f32`. (Covered by ODS).
I7: a) `feature_index` is not a constant of type `si64`. (Covered by ODS).
C1: a) `feature_index` < 0.
    b) `feature_index` >= rank(`operand`).
C2: a) element_type(`operand`) != element_type(`scale`). (Covered by ODS).
    b) element_type(`operand`) != element_type(`mean`). (Covered by ODS).
    c) element_type(`operand`) != element_type(`variance`). (Covered by ODS).
    d) element_type(`operand`) != element_type(`grad_output`). (Covered by ODS).
    e) element_type(`operand`) != element_type(`grad_operand`). (Covered by ODS).
    f) element_type(`operand`) != element_type(`grad_scale`). (Covered by ODS).
    g) element_type(`operand`) != element_type(`grad_offset`). (Covered by ODS).
C3: a) shape(`operand`) != shape(`grad_output`).
    b) shape(`operand`) != shape(`grad_operand`).
C4: a) shape(`scale`) != shape(`mean`).
    b) shape(`scale`) != shape(`variance`).
    c) shape(`scale`) != shape(`grad_scale`).
    d) shape(`scale`) != shape(`grad_offset`).
C5: a) size(`scale`) != dim(operand, feature_index).
```

If we drop the "Covered by ODS" pieces, this will leave us with the following test cases:

```
C1a: `feature_index` < 0.
C1b: `feature_index` >= rank(`operand`).
C3a: shape(`operand`) != shape(`grad_output`).
C3b: shape(`operand`) != shape(`grad_operand`).
C4a: shape(`scale`) != shape(`mean`).
C4b: shape(`scale`) != shape(`variance`).
C4c: shape(`scale`) != shape(`grad_scale`).
C4d: shape(`scale`) != shape(`grad_offset`).
C5a: size(`scale`) != dim(operand, feature_index).
```

Notes:
* Added `i6` in the spec to better align with the spec comments referring to [batchnorm_expander.cc](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/service/batchnorm_expander.cc#L521-L524) in XLA.
* `size` -> `num_elements` to be consistent with what's written in `compute_mean` function.

closes #1121

### CODE REVIEW

**Comment by ghpvnist** [comment]
Please review the latest [commit](https://github.com/openxla/stablehlo/pull/1394/commits/031c95c2bd7d27c57d8c6d7416dc5541b71c2fae) only.

File: stablehlo/reference/Ops.cpp [hunk 1]
```diff
@@ -381,6 +489,19 @@ Tensor evalBatchNormInferenceOp(const Tensor &operand, const Tensor &scale,
...
+SmallVector<Tensor> evalBatchNormTraining(const Tensor &operand,
+                                          const Tensor &scale,
+                                          const Tensor &offset, APFloat epsilon,
+                                          const Axis featureIndex,
+                                          ArrayRef<ShapedType> resultTypes) {
+  auto mean = computeMean(operand, featureIndex);
+  auto variance = computeVariance(operand, featureIndex);
+  return {evalBatchNormInferenceOp(operand, scale, offset, mean, variance,
+                                   epsilon, featureIndex,
+                                   resultTypes[0].dyn_cast<ShapedType>()),
```
Comment (burmako) [inline stablehlo/reference/Ops.cpp:620]: Doesn't look like the cast is needed here.

File: stablehlo/reference/Ops.cpp [hunk 2]
```diff
@@ -381,6 +489,19 @@ Tensor evalBatchNormInferenceOp(const Tensor &operand, const Tensor &scale,
...
+          mean, variance};
```
Comment (burmako) [inline stablehlo/reference/Ops.cpp:621]: Nice catch! Let's update the spec to return multiple results too.

File: stablehlo/reference/Ops.cpp [hunk 3]
```diff
@@ -30,13 +33,106 @@ namespace mlir {
 namespace stablehlo {
 namespace {
 
+static Type getF64Type(MLIRContext *context) {
+  return mlir::Builder(context).getF64Type();
+}
+
+static Type getI64Type(MLIRContext *context) {
+  return mlir::Builder(context).getI64Type();
+}
```
Comment (burmako) [inline stablehlo/reference/Ops.cpp:42]: Given that these functions are used only once, let's inline them into their callsites?

Comment (burmako) [inline stablehlo/reference/Ops.cpp:42]: Also, similarly to the nearby code reviews, let's use the builderless formulations via FloatType::get and IntegerType::get - when builders are not available in the lexical context, they are shorter to type and more idiomatic.

File: stablehlo/reference/Ops.cpp [hunk 4]
```diff
@@ -30,13 +33,106 @@ namespace mlir {
...
+      {}, {operand.getType()}, {RankedTensorType::get({}, f64Type)},
+      getDenseIntElementsAttr(i64Type, dimensions, {}), inferredReduceType);
+
+  if (failed(reduceStatus))
+    report_fatal_error(
+        invalidArgument("Could not infer ReduceOp's return type"));
+
+  auto reducedSum = computeSum(
+      operand,
+      Tensor(RankedTensorType::get({}, f64Type), Element(f64Type, 0.0)),
```
Comment (burmako) [inline stablehlo/reference/Ops.cpp:110]: Why are we using `f64` here? Shouldn't this be `operand.getElementType()`?

Comment (burmako) [inline stablehlo/reference/Ops.cpp:110]: Also, the spec needs to be updated to properly express this. Instead of `0.0`, we'll probably want to say `constant(0.0, type(operand))`, and instead of `constant(...)` a few lines below, we'll probably want to say `constant(..., type(operand))`?

File: stablehlo/reference/Ops.cpp [hunk 5]
```diff
@@ -30,13 +33,106 @@ namespace mlir {
...
+  dimensions.erase(dimensions.begin() + featureIndex);
+
+  SmallVector<ShapedTypeComponents> inferredReduceType;
+  auto reduceStatus = hlo::inferReduceOp(
+      {}, {operand.getType()}, {RankedTensorType::get({}, f64Type)},
+      getDenseIntElementsAttr(i64Type, dimensions, {}), inferredReduceType);
+
+  if (failed(reduceStatus))
+    report_fatal_error(
+        invalidArgument("Could not infer ReduceOp's return type"));
```
Comment (burmako) [inline stablehlo/reference/Ops.cpp:106]: Wow this is really a mouthful - I'll try to find time tomorrow to think how to make this look better. Sorry I didn't have the time to look into this in the last few weeks.

File: stablehlo/reference/Ops.cpp [hunk 6]
```diff
@@ -30,13 +33,106 @@ namespace mlir {
...
+      operand,
+      Tensor(RankedTensorType::get({}, f64Type), Element(f64Type, 0.0)),
+      Axes(dimensions),
+      RankedTensorType::get(inferredReduceType[0].getDims(),
+                            operand.getElementType()));
+
+  auto divisor = Tensor(
+      RankedTensorType::get({}, f64Type),
+      Element(f64Type, static_cast<double>(operand.getNumElements() /
+                                           operand.getShape()[featureIndex])));
```
Comment (burmako) [inline stablehlo/reference/Ops.cpp:118]: Same comment here - why are we using `f64` and not `operand.getElementType()`?

File: stablehlo/reference/Ops.h [hunk 7]
```diff
@@ -34,6 +34,11 @@ Tensor evalBatchNormInferenceOp(const Tensor &operand, const Tensor &scale,
                                 const Tensor &offset, const Tensor &mean,
                                 const Tensor &variance, APFloat epsilon,
                                 Axis featureIndex, ShapedType resultType);
+SmallVector<Tensor> evalBatchNormTraining(const Tensor &operand,
+                                          const Tensor &scale,
+                                          const Tensor &offset, APFloat epsilon,
+                                          const Axis featureIndex,
```
Comment (burmako) [inline stablehlo/reference/Ops.h:46]: I think the `const` part in `const Axis` is a bit over the top. Are we trying to guard against `evalBatchNormTraining` modifying the `featureIndex` parameter inside the body?

**Review by burmako — CHANGES_REQUESTED** [review]

File: docs/spec.md [hunk 8]
```diff
@@ -949,23 +949,25 @@ def batch_norm_grad(operand, scale, mean, variance, grad_output, epsilon, featur
...
   # Temporary variables have exactly the same names as in the C++ code
   elements_per_feature = constant(
-    divide(size(operand), dim(operand, feature_index)))
+    divide(num_elements(operand), dim(operand, feature_index)))
   i1 = multiply(
     grad_output,
     broadcast_in_dim(elements_per_feature, [], shape(operand)))
   i2 = broadcast_in_dim(
-    compute_sum(grad_output, feature_index),
+    compute_sum(grad_output),
```
Comment (burmako) [inline docs/spec.md:957]: Not sure I understand the reasoning behind this change.

File: stablehlo/reference/Ops.cpp [hunk 9]
```diff
@@ -459,6 +494,90 @@ Tensor evalAndOp(const Tensor &lhs, const Tensor &rhs, ShapedType resultType) {
...
+                                         operand.getShape()[featureIndex])));
+
+  auto i1 = evalMultiplyOp(
+      gradOutput,
+      evalBroadcastInDimOp(elementsPerFeature, {}, operand.getType()),
+      gradOutput.getType());
+
+  auto f64Type = getF64Type(operand.getType().getContext());
+  auto initValue =
+      Tensor(RankedTensorType::get({}, f64Type), Element(f64Type, 0.0));
```
Comment (burmako) [inline stablehlo/reference/Ops.cpp:535]: Same question as for batch_norm_training - why are we using f64 here?

**Review by burmako — CHANGES_REQUESTED** [review]

**Comment by burmako** [comment]
Apologies - I'm realizing that I've reviewed both batch_norm_training and batch_norm_grad in this PR.

File: docs/spec.md (hunk 8 above)
Comment (ghpvnist) [reply docs/spec.md:957]: Not sure what happened either. I meant to add `feature_index` for `i3`.

File: stablehlo/tests/ops_stablehlo.mlir [hunk 10]
```diff
@@ -4754,48 +4754,64 @@ func.func @error_incompatible_alias_element_types (%arg0: tensor<2xf32> {stableh
...
 }
 
 // -----
 
-func.func @error_batch_norm_train(%input: tensor<2x2x2x2xf32>, %scale: tensor<3xf32>, %offset: tensor<3xf32>) -> tensor<2x2x2x2xf32> {
+func.func @batch_norm_training_c3_c4(%input: tensor<2x2x2x2xf32>, %scale: tensor<3xf32>, %offset: tensor<3xf32>) -> tensor<2x2x2x2xf32> {
   // expected-error@+2 {{failed to infer returned types}}
   // expected-error@+1 {{expects the size of single-dimensional operands to be compatible with feature count, but the size of single-dimensional operands is 3 and the feature count is 2.}}
-  %0:3 = "stablehlo.batch_norm_training" (%input, %scale, %offset) {epsilon = 0.001 : f32, feature_index = 3 : i64} : (tensor<2x2x2x2xf32>, tensor<3xf32>, tensor<3xf32>) -> (tensor<2x2x2x2xf32>, tensor<3xf32>, tensor<3xf32>)
+  %0:3 = "stablehlo.batch_norm_training" (%input, %scale, %offset) {
```
Comment (sdasgup3) [inline stablehlo/tests/ops_stablehlo.mlir:4810]: In other similar examples in the current file, we do not have the multiline format (like like 4900). Can we make it uniform throughout

File: stablehlo/tests/infer_stablehlo.mlir [hunk 11]
```diff
@@ -256,7 +256,11 @@ func.func @fft(%arg0: tensor<3x9xcomplex<f32>>) -> tensor<3x9xindex> {
 
 // CHECK-LABEL: func @batch_norm_grad
 func.func @batch_norm_grad(%input: tensor<2x2x2x2xf32>, %scale: tensor<2xf32>, %mean: tensor<2xf32>, %variance: tensor<2xf32>, %grad_output: tensor<2x2x2x2xf32>) -> tensor<*xindex> {
-  %0:3 = "stablehlo.batch_norm_grad" (%input, %scale, %mean, %variance, %grad_output) {epsilon = 0.001 : f32, feature_index = 0 : i64} : (tensor<2x2x2x2xf32>, tensor<2xf32>, tensor<2xf32>, tensor<2xf32>, tensor<2x2x2x2xf32>) -> (tensor<2x2x2x2xf32>, tensor<2xf32>, tensor<2xf32>)
+  %0:3 = "stablehlo.batch_norm_grad" (%input, %scale, %mean, %variance, %grad_output) {
```
Comment (sdasgup3) [inline stablehlo/tests/infer_stablehlo.mlir:259]: Same comment about changing the format.

File: stablehlo/tests/infer_stablehlo.mlir [hunk 12]
```diff
@@ -336,28 +344,60 @@ func.func @batch_norm_grad_bounds(
 
 // -----
 
-// CHECK-LABEL: func @batch_norm_train_bounds
-func.func @batch_norm_train_bounds(
-  %input: tensor<2x?xf32, #stablehlo.bounds<?, 64>>,
```
Comment (sdasgup3) [inline stablehlo/tests/infer_stablehlo.mlir:341]: It seems we have deleted this?

**Review by sdasgup3 — COMMENTED** [review]
reviewed the pieces related to batch_norm_grad op. I believe batch_norm_training is reviewed separately (https://github.com/openxla/stablehlo/pull/1393).

File: stablehlo/tests/ops_stablehlo.mlir (hunk 10 above)
Comment (ghpvnist) [reply stablehlo/tests/ops_stablehlo.mlir:4810]: Unfortunately the examples are inconsistent throughout the file. Some tests like [pad](https://github.com/openxla/stablehlo/blob/main/stablehlo/tests/ops_stablehlo.mlir#L5329-L5337) do have formatted tests, and the recently merged [batch_norm_inference](https://github.com/openxla/stablehlo/blob/main/stablehlo/tests/ops_stablehlo.mlir#L4806-L4813) is also formatted. Since formatting provides more readability, I'd say let's keep it?

File: stablehlo/tests/infer_stablehlo.mlir (hunk 12 above)
Comment (ghpvnist) [reply stablehlo/tests/infer_stablehlo.mlir:341]: Thanks, added back in BatchNormTraining PR

File: stablehlo/tests/infer_stablehlo.mlir (hunk 11 above)
Comment (ghpvnist) [reply stablehlo/tests/infer_stablehlo.mlir:259]: See above.

**Review by sdasgup3 — APPROVED** [review]

File: stablehlo/reference/Ops.cpp [hunk 13]
```diff
@@ -590,6 +600,76 @@ Tensor evalAtan2Op(const Tensor &lhs, const Tensor &rhs,
...
+  auto scaleBroadcast =
+      evalBroadcastInDimOp(scale, {featureIndex}, operand.getType());
+  auto meanBroadcast =
+      evalBroadcastInDimOp(mean, {featureIndex}, operand.getType());
+  auto varianceBroadcast =
+      evalBroadcastInDimOp(variance, {featureIndex}, operand.getType());
+  auto epsilonBroadcast = evalBroadcastInDimOp(
+      makeTensor(DenseElementsAttr::get(
+          RankedTensorType::get({}, operand.getElementType()), {epsilon})),
+      {}, operand.getType());
```
Comment (burmako) [inline stablehlo/reference/Ops.cpp:618]: `constant(epsilon)` within `epsilon_bcast = broadcast_in_dim(constant(epsilon), [], shape(operand))` needs to be updated accordingly.

File: docs/spec.md [hunk 14]
```diff
@@ -1048,23 +1048,25 @@ def batch_norm_grad(operand, scale, mean, variance, grad_output, epsilon, featur
   # Use the implementation from batchnorm_expander.cc in XLA
   # Temporary variables have exactly the same names as in the C++ code
   elements_per_feature = constant(
-    divide(size(operand), dim(operand, feature_index)))
+    divide(num_elements(operand), dim(operand, feature_index)))
```
Comment (burmako) [inline docs/spec.md:1051]: This needs an explicit type too.

File: docs/spec.md [hunk 15]
```diff
@@ -1048,23 +1048,25 @@ def batch_norm_grad(operand, scale, mean, variance, grad_output, epsilon, featur
...
     [feature_index], shape(operand))
   i3 = broadcast_in_dim(
-    compute_sum(multiply(grad_output, centered_operand)),
+    compute_sum(multiply(grad_output, centered_operand), feature_index),
     [feature_index], shape(operand))
   i4 = multiply(i3, centered_operand)
   i5 = divide(i4, add(variance_bcast, epsilon_bcast))
+  i6 = subtract(subtract(i1, i2), i5)
   grad_operand = multiply(
     divide(divide(scale_bcast, stddev), elements_per_feature),
```
Comment (burmako) [inline docs/spec.md:1065]: Looks like elements_per_feature needs to be broadcast. Let's do that at the definition site?

**Review by burmako — CHANGES_REQUESTED** [review]

**Review by burmako — APPROVED** [review]
